### ISSUE TITLE
Game is becoming slow after long flights

### ISSUE THREAD

**Issue by Mc-Pain** [pioneerspacesim/pioneer#5735]
Loading this savefile will bring the game into slowness.

[_quicksave.gz](https://github.com/pioneerspacesim/pioneer/files/14150868/_quicksave.gz)

I suspect it is being accululated through gameplay until next jump.

Will investigate further

**Comment by Web-eWorks** [comment]
@Mc-Pain ...have you rebuilt your model cache? `make -C build build-data` - I had this issue after the manual docking PR was merged, as ships built up at the edge of stations and couldn't land because of collision issues with out-of-date models.

**Comment by Mc-Pain** [comment]
> @Mc-Pain ...have you rebuilt your model cache? `make -C build build-data`

Not yet, rebuilding from scratch

**Comment by Mc-Pain** [comment]
Rebuilt from scratch, got this:

[pioneer-profile-20240204_022523.html.gz](https://github.com/pioneerspacesim/pioneer/files/14154529/pioneer-profile-20240204_022523.html.gz)

**Comment by Web-eWorks** [comment]
So - this is not a significant issue when compiling in release configuration, on my hardware at least. It does result in some choppy frame timing with vsync, and will occasionally cause 33ms frames (dipping to an average of 45fps in the worst case).

The root of the issue that I have been able to discover comes from the fact that the tradeship count at Sol starts at ~450 and after 30 days of timewarp has reached 770 total ships, of which ~400 are "active" and flying through space. Because ship computation cost is super-linear (i.e. `O(n log n)`), this results in markedly worse performance, given that over twice as many ships are currently flying around.

In a release build, the performance loss is negligible enough (and most player's playstyles do *not* involve repeatedly spending 30 consecutive days in a single system) that I don't think this is of significant severity to warrant a hotfix. It is definitely something I'd like to address over time however.

The profile you've provided seems indicative that some issue is preventing ships from being able to dock at the Mao's Grave station; there are approximately 1731 ship-object collision pairs happening there per physics tick. This is *strongly* indicative of outdated cached model files; all station and landing pad models were updated during the last release and if you had run the model compiler at any point you will have outdated/invalid collision data which will prevent **any** ships from landing at those stations.

**Comment by impaktor** [comment]
> Sol starts at ~450 and after 30 days of timewarp has reached 770 total ships.

Ideally each system should be spawned in equilibrium, so there will be no significant change in total number of trade ships over time.

**Comment by Mc-Pain** [comment]
After models recompilation:
![image](https://github.com/pioneerspacesim/pioneer/assets/9453287/96309c82-f98a-4d07-985f-3c90b80d6b37)
[pioneer-profile-20240204_175057.html.gz](https://github.com/pioneerspacesim/pioneer/files/14157238/pioneer-profile-20240204_175057.html.gz)

**Comment by nozmajner** [comment]
I've tested this at the Cydonia start, and seem to get a performance drop quite soon. In 13 minutes, it is at about 13 FPS. OBS runs in the background, but that doesn't seem to affect it much.
Tradeship count goes from around 250 to 750 during this quarter hour.
If I disable the Tradeships module altogether, it doesn't happen. Nor any noticeable fluctuations in the FPS.

[Here's the recording of it](https://www.youtube.com/watch?v=pOZfGvVUrbU)

[Output.txt of this recording](https://github.com/pioneerspacesim/pioneer/files/14165208/output.txt)

CPU thread load looks like this:
![image](https://github.com/pioneerspacesim/pioneer/assets/4182678/a7a7fa00-3e21-44d8-bd96-48a2f517ba5c)
CPU data sheet:
![image](https://github.com/pioneerspacesim/pioneer/assets/4182678/0e6cad4c-4a0e-4d33-9759-edfab9524469)

**Comment by nozmajner** [comment]
I'm noticing the same in the previous release too. Not exactly with the same amount of ships, and sometimes the FPS fluctuates back to higher values.
![image](https://github.com/pioneerspacesim/pioneer/assets/4182678/c16535ee-c86f-4623-a1cc-754eaa4a6c4b)

**Comment by Gliese852** [comment]
I think the bulk of the tradeships that accumulate are those trying to land on a planet that don't have powerful enough thrusters. 
This is not yet [pre-checked](https://github.com/pioneerspacesim/pioneer/blob/3d4553fd9367d9ff11cf31535eebaba7760f6856/data/modules/TradeShips/Trader.lua#L217).

The mechanism is this: if there is a free pad for the ship in the port, it tries to land and crashes.
If there is no pad available, ship goes into the orbit of the planet and waits there for 6 hours. After this, `AICmdDock` starts, when the autopilot starts, it [turns out](https://github.com/pioneerspacesim/pioneer/blob/3d4553fd9367d9ff11cf31535eebaba7760f6856/src/ShipAICmd.cpp#L1295) that the gravity is too high, the autopilot turns off and the tradeship turns into a zombie. This is visible in the log:
![изображение](https://github.com/pioneerspacesim/pioneer/assets/18342621/dedcdf63-ce2f-4d28-8ade-3fb483334b52)

I added a pre-check and now the ships don't accumulate, you can see that after waiting (error `dock_aft_6h`) all ships are still alive (or docked, or jumped out, or autopilot active):
![изображение](https://github.com/pioneerspacesim/pioneer/assets/18342621/e1486c0d-5abb-4ad1-bed4-0c9d78964f6d)
I'll PR this fix soon.

**Comment by Gliese852** [comment]
What did I find!
After I eliminated all the pathological accumulations of tradeships, still, after maybe a week in-game, fps dropped significantly.
Apparently this is why:
![изображение](https://github.com/pioneerspacesim/pioneer/assets/18342621/772f77c7-bbea-48b4-b437-6e56ba461d07)
It can be seen that one "timestep" has become longer by ~12 MCycles, ~10 of which are `LuaTimer::Tick()`.

I looked at this function in the debugger - there are about a thousand callbacks that need to be called every 5 seconds, but each of them is called every frame. Not only called, but removed from the queue, inserted into the array, then reinserted into the queue. Because after calling the callback, _only_ 5 seconds are added to its 'call at' time, which means that after a few seconds, with time accelerated by at least 10x, they become hopelessly outdated and begin to fire every frame.
P.S. Callbacks are most likely from here: https://github.com/pioneerspacesim/pioneer/blob/a7f2781dc9ddcfda6a21fad4239dbec13aff5621/data/modules/LifeSupport.lua#L51
